config/settings.yml

```diff
@@ -0,0 +1,15 @@
+# Beware! This file is public by definition
+default: &default
+  application_name: 'u-can-act'
+  logo:
+    mentor_logo: 'U_can_act_logo_CMYK_BLAUW.png'
+    student_logo: 'U_can_act_logo_CMYK_GROEN.png'
+    fallback_logo: 'U_can_act_logo_ZWART.png'
+development:
+  <<: *default
+production:
+  <<: *default
+test:
+  <<: *default
+staging:
+  <<: *default
```

Comment:
Ik snap niet precies waarom projectspecifieke settings (dus dingen die verschillen tussen vsv, en andere projecten), waarom die zouden verschillen per environment. Is het niet juist alleen de ENV vars (dus passwords, tokens, e.d. die kunnen verschillen tussen de verschillende environments? En zijn dat juist dingen die je niet in een public config file wilt zetten?

Projectspecifieke settings ben ik helemaal voor, maar ik zat dan meer te denken in de zin van:
- We hebben een map ergens in de repo die we "deployments" oid noemen.
- Die map heeft heeft submappen voor iedere "klant"/project. vsv, etc.
- Er is een default map genaamd default, of template, oid, van dat is een default organisatie met minimale settings.
- Het idee is dat alle projectspecifieke dingen geisoleerd worden tot die map, en als we een nieuw project toevoegen, dat we dan de default map kopieren, renamen, en daar een paar dingen in aanpassen.
- Er is een database setting die zegt welke map (welk project/deployment) de huidige is.

Wat staat er in de map? Ik zat te denken aan een structuur van 
  assets/css/project.scss - projectspecifieke css (eventueel zou je deze ook kunnen opsplitsen in een mentor.scss en een student.scss)
   assets/javascript/project.js - projectspecifieke javascriptcode

- En dan laden we in application.js dus deployments/$PROJECT/assets/javascripts/project.js en in application.scss laden includen we deployments/$PROJECT/assets/css/project.scss (je snapt het wel).
- Dus in plaats van dat we hierboven zeggen oh voor het mentor logo moet je deze bestandsnaam hebben, zeggen we in de mentor.scss van oh deze css klasse definieer ik de background met dit image url, en de bijbehorende image zelf staat dan in dezelfde map deployments/$PROJECT/assets/images/mentor_logo.png oid.

Voor teksten in de app zouden we dan i18n moeten gebruiken, waarbij we ook locales laden uit de huidge deployment map (dus er staat bv ook een nl.yml in iedere deployment map), en dan kunnen we makkelijk alle teksten in de app aanpassen en daar zouden we eventueel ook de projectnaam in kunnen gooien. Dus dat we die gewoon met i18n doen ipv daar een aparte setting voor te maken. Dan is het makkelijk dat alle teksten, alles dat als tekst voorkomt in de app dat dat gewoon i18n is.

En de deployments mappen hebben ook een settings.yml file zoals hier, maar ik stel voor dat we die voornamelijk voor "feature toggles" gebruiken. Dwz, van voor sommige organisaties willen we misschien geen admin dashboard en dan hebben dus een enable_admindashboard flag die dan op true of false kan worden gezet. Of een boolean flag van "use random sms texts". Het is er dus meer voor bedoeld dat we bepaalde functionaliteit aan- of uit kunnen zetten per project. Of als we een nieuwe functionaliteit toevoegen dat die dan achter zo'n feature flag komt te hangen en dat we die eerst alleen bij een bepaald project uitproberen, dat soort dingen. Misschien hebben we hier iets meer nodig dan simpele .yml files, het zou namelijk cool zijn als iedere flag een default waarde heeft zodat we per project alleen de flags die niet default zijn hoeven te specificeren oid.


Om te beginnen zouden we denk ik alle strings moeten i18nnen, zo'n deployments map moeten aanmaken, en alle css en alle strings waarvan we zoiets hebben van dit zouden verschillende projecten misschien anders willen tonen, dat we die code overhevelen naar de deployments/default map (of voor erg vsv-specifieke dingen, dat we ze in de deployments/vsv map zetten en dan een "neutralere variant"  in de deployments/default map zetten).

Reply:
> Ik snap niet precies waarom projectspecifieke settings (dus dingen die verschillen tussen vsv, en andere projecten), waarom die zouden verschillen per environment. Is het niet juist alleen de ENV vars (dus passwords, tokens, e.d. die kunnen verschillen tussen de verschillende environments? En zijn dat juist dingen die je niet in een public config file wilt zetten?

Ja dat leek mij ook niet per se nodig, maar kan af en toe ook wel handig zijn denk ik. In ieder geval, we hoeven geen dingen te copy pasten omdat we gewoon dat defaults block kunnen hergebruiken. Wat ik wou met dit systeem is eigenlijk in ieder geval geen externe gem gebruiken als deze oplossing nagenoeg hetzelfde doet. Er bestaan een hoop gems voor configuratie, maar dat lijkt mij onnodig, en dit werkt volgens mij goed genoeg. Deze file is in ieder geval niet bedoelt voor secrets, maar eigenlijk is mijn idee om deze file (op een template file na) helemaal niet in deze repo te zetten.

> Projectspecifieke settings ben ik helemaal voor, maar ik zat dan meer te denken in de zin van:
> - We hebben een map ergens in de repo die we "deployments" oid noemen.
> - Die map heeft heeft submappen voor iedere "klant"/project. vsv, etc.
> - Er is een default map genaamd default, of template, oid, van dat is een default organisatie met minimale settings.
> - Het idee is dat alle projectspecifieke dingen geisoleerd worden tot die map, en als we een nieuw project toevoegen, dat we dan de default map kopieren, renamen, en daar een paar dingen in aanpassen.
> - Er is een database setting die zegt welke map (welk project/deployment) de huidige is.
> 
> Wat staat er in de map? Ik zat te denken aan een structuur van 
>   assets/css/project.scss - projectspecifieke css (eventueel zou je deze ook kunnen opsplitsen in een mentor.scss en een student.scss)
>    assets/javascript/project.js - projectspecifieke javascriptcode
>
> - En dan laden we in application.js dus deployments/$PROJECT/assets/javascripts/project.js en in application.scss laden includen we deployments/$PROJECT/assets/css/project.scss (je snapt het wel).
> - Dus in plaats van dat we hierboven zeggen oh voor het mentor logo moet je deze bestandsnaam hebben, zeggen we in de mentor.scss van oh deze css klasse definieer ik de background met dit image url, en de bijbehorende image zelf staat dan in dezelfde map deployments/$PROJECT/assets/images/mentor_logo.png oid.

Ja dat is ook waar ik aan zat te denken, maar dan net even wat anders. Eigenlijk vind ik het niet mooi als we onze repository vervuilen met al die project specifieke crap. En dan bedoel ik dus dat we niet voor elk project iets zouden moeten toevoegen aan deze repo. Ik zat daarom zelf meer te denken aan een soort assets achtige repo / S3 oplossing, waar elk van de project specifieke projecten een kleine set met assets heeft en we die dan op de een of andere manier mounten in deze app. Ik weet nog niet helemaal wat daar de beste manier voor is maar er zijn verschillende opties mogelijk:
- We kunnen per subproject een nieuwe repo maken als fork van deze en daar dingen in aanpassen (__dit willen we sowieso proberen te vermijden__)
- We maken een repo per project waar we dit project in mounten als soort engine (__ook liever niet denk ik__)
- We maken per project een repo met daarin alleen de assets (of we slaan ze ergens anders in op), en maken een soort capistrano setup die alles samenvoegt (__ik denk dat zoiets het beste is tbh__)

> Voor teksten in de app zouden we dan i18n moeten gebruiken, waarbij we ook locales laden uit de huidge deployment map (dus er staat bv ook een nl.yml in iedere deployment map), en dan kunnen we makkelijk alle teksten in de app aanpassen en daar zouden we eventueel ook de projectnaam in kunnen gooien. Dus dat we die gewoon met i18n doen ipv daar een aparte setting voor te maken. Dan is het makkelijk dat alle teksten, alles dat als tekst voorkomt in de app dat dat gewoon i18n is.

Ja dat kan inderdaad, maar ook daar geldt weer hetzelfde voor lijkt me. In deze app zetten we een template bestand, en in alle andere projecten (waar u-can-act dan eigenlijk ook onder valt) zetten we de specifieke implementatie van het i18n bestand.

> En de deployments mappen hebben ook een settings.yml file zoals hier, maar ik stel voor dat we die voornamelijk voor "feature toggles" gebruiken. Dwz, van voor sommige organisaties willen we misschien geen admin dashboard en dan hebben dus een enable_admindashboard flag die dan op true of false kan worden gezet. Of een boolean flag van "use random sms texts". Het is er dus meer voor bedoeld dat we bepaalde functionaliteit aan- of uit kunnen zetten per project. Of als we een nieuwe functionaliteit toevoegen dat die dan achter zo'n feature flag komt te hangen en dat we die eerst alleen bij een bepaald project uitproberen, dat soort dingen. Misschien hebben we hier iets meer nodig dan simpele .yml files, het zou namelijk cool zijn als iedere flag een default waarde heeft zodat we per project alleen de flags die niet default zijn hoeven te specificeren oid.

Ja dat kan hier ook inderdaad bij in, maar ik denk niet dat we het daarbij hoeven te houden. Ik denk dat we moeten proberen onze .ENV niet langer te vervuilen met dingen die ook best in zo'n config bestand mogen staan. Dan kunnen we namelijk gewoon de .ENV blijven gebruiken voor secrets, en alles wat daar niet binnen valt op nemen in zo'n config / settings yaml. Denk bijvoorbeeld ook aan 'announcement weeks' en whatever, wat niet echt i18n is, maar ook geen ENV.

> Om te beginnen zouden we denk ik alle strings moeten i18nnen, zo'n deployments map moeten aanmaken, en alle css en alle strings waarvan we zoiets hebben van dit zouden verschillende projecten misschien anders willen tonen, dat we die code overhevelen naar de deployments/default map (of voor erg vsv-specifieke dingen, dat we ze in de deployments/vsv map zetten en dan een "neutralere variant"  in de deployments/default map zetten).

Dat is inderdaad de volgende stap. Met dit PR wou ik in ieder geval de mogelijkheid toevoegen dat we configuratie ook in een apart bestand kunnen zetten, waar dat nu misschien nog ge hardcode is. Als we voor u-can-act dit allemaal netjes in een config bestand hebben gezet kunnen we die eruit gaan trekken en dit config bestand om gaan zetten naar een template bestand.

Dus tl;dr, waar ik aan zit te denken is het volgende:
- Elk project heeft een settings.yml (zoals deze).
- Bij een deploy doen we (hoe precies weet ik nog niet) een checkout van deze repo, en vervangen deze settings.yml door de project specifieke settings.yml
- Dan voegen we de project specifieke seeds toe
- We voegen de assets toe die van belang zijn voor dat specifieke project
- Dan, als alle project specifieke dingen erin staan moeten we misschien de integratie specs gaan draaien (of alle specs, kan ook), om te testen of alle assets e.d. bestaan / alles wat nodig is om de app te draaien aanwezig is. Hier moeten we goed over na denken, want dit kan snel te veel worden om handmatig te managen.
- Als die tests goed gaan pushen we alles naar de server.

En dit moet allemaal gebeuren in CI / CD, dus ik denk dat we per repo een soort CI moeten hebben dat automatisch al deze dingen doet. Mochten we dan een upgrade doen aan vsv zelf dan zouden we er nog een scriptje aan kunnen hangen dat dan een auto deploy doet van alle losse projecten.

Reply:
https://freedcamp.com/Compsy_oK1/VSV_wNL/todos/18282841/